source/common/http/conn_manager_impl.cc

```diff
@@ -1550,21 +1563,50 @@ void ConnectionManagerImpl::ActiveStream::refreshCachedTracingCustomTags() {
   }
 }
 
+// TODO(chaoqin-li1123): Make on demand vhds and on demand srds works at the same time.
 void ConnectionManagerImpl::ActiveStream::requestRouteConfigUpdate(
-    Event::Dispatcher& thread_local_dispatcher,
     Http::RouteConfigUpdatedCallbackSharedPtr route_config_updated_cb) {
-  ASSERT(!filter_manager_.requestHeaders()->Host()->value().empty());
-  const auto& host_header = absl::AsciiStrToLower(filter_manager_.requestHeaders()->getHostValue());
-  route_config_update_requester_->requestRouteConfigUpdate(host_header, thread_local_dispatcher,
-                                                           std::move(route_config_updated_cb));
+  absl::optional<Router::ConfigConstSharedPtr> route_config = routeConfig();
+  Event::Dispatcher& thread_local_dispatcher =
+      connection_manager_.read_callbacks_->connection().dispatcher();
+  if (route_config.has_value() && route_config.value()->usesVhds()) {
+    // On demand vhds
+    ASSERT(!filter_manager_.requestHeaders()->Host()->value().empty());
+    const auto& host_header =
+        absl::AsciiStrToLower(filter_manager_.requestHeaders()->getHostValue());
+    route_config_update_requester_->requestVhdsUpdate(host_header, thread_local_dispatcher,
+                                                      std::move(route_config_updated_cb));
+    return;
+  } else if (snapped_scoped_routes_config_ != nullptr &&
+             (scope_key_hash_ = snapped_scoped_routes_config_->computeKeyHash(
```

Comment:
hash is a scopeKey internal impl detail, do we want to keep it in side scope RDS? on the other hand,whynot return the scope key instead here?

Reply:
It doesn't make much difference, we still need to convert the scope key to its hash later if we return a scope key here.

Reply:
you move the scopeKey with a shard_ptr bound to the callback, I dont think that's very expensive. With that, Activestream doesnt need to know about the hash() of a ScopeKey, which is kinda internal.

Reply:
I think this feedback is consistent with the above comment by @dmitri-d. Ideally we minimize the amount of on-demand (S)RDS/VHDS business logic we are baking into HCM.

Reply:
I will try to move the logic into the requester.